src/network/stream_reader.rs

```diff
@@ -0,0 +1,310 @@
+    }
+
+    #[test]
+    fn read_singlemsg_test() {
+        let mut stream = init_stream(&MSG_VERSION);
+        let message = StreamReader::new(&mut stream, None).next_message().unwrap();
+
+        check_version_msg(&message);
+    }
+
+    #[test]
+    fn read_doublemsgs_test() {
+        let mut stream = init_stream(&MSG_VERSION);
+        write_file(&mut stream, &MSG_PING);
+
+        let mut reader = StreamReader::new(&mut stream, None);
+        let message = reader.next_message().unwrap();
+        check_version_msg(&message);
+
+        let msg = reader.next_message().unwrap();
+        assert_eq!(msg.magic, 0xd9b4bef9);
+        if let NetworkMessage::Ping(nonce) = msg.payload {
+            assert_eq!(nonce, 100);
+        } else {
+            panic!("Wrong message type, expected PingMessage");
+        }
+    }
+
+    fn serve_tcp(port: u16, pieces: Vec<Vec<u8>>) {
+        let listener = TcpListener::bind(format!("127.0.0.1:{}", port)).unwrap();
+        for ostream in listener.incoming() {
+            let mut ostream = ostream.unwrap();
+
+            for piece in pieces {
+                ostream.write(&piece[..]).unwrap();
+                ostream.flush().unwrap();
+                thread::sleep(Duration::from_secs(1));
+            }
+
+            ostream.shutdown(Shutdown::Both).unwrap();
+            break;
+        }
+    }
+
+    #[test]
+    fn read_multipartmsg_test() {
+        let port: u16 = 34254;
+        let handle = thread::spawn(move || {
+            serve_tcp(port, vec![MSG_VERSION[..24].to_vec(), MSG_VERSION[24..].to_vec()]);
+        });
+
+        thread::sleep(Duration::from_secs(1));
+        let mut istream = TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
+        let message = StreamReader::new(&mut istream, None).next_message().unwrap();
+        check_version_msg(&message);
+
+        handle.join().unwrap();
+    }
+
+    #[test]
+    fn read_sequencemsg_test() {
+        let port: u16 = 34255;
+        let handle = thread::spawn(move || {
+            serve_tcp(port, vec![
+                // Real-world Bitcoin core communication case for /Satoshi:0.17.1/
+                MSG_VERSION[..23].to_vec(), MSG_VERSION[23..].to_vec(),
+                MSG_VERACK.to_vec(),
+                MSG_ALERT[..24].to_vec(), MSG_ALERT[24..].to_vec()
+            ]);
+        });
+
+        thread::sleep(Duration::from_secs(1));
+        let mut istream = TcpStream::connect(format!("127.0.0.1:{}", port)).unwrap();
+
+        let mut reader = StreamReader::new(&mut istream, None);
+        let message = reader.next_message().unwrap();
+        check_version_msg(&message);
+
+        let msg = reader.next_message().unwrap();
+        assert_eq!(msg.magic, 0xd9b4bef9);
+        assert_eq!(msg.payload, NetworkMessage::Verack, "Wrong message type, expected PingMessage");
+
+        let msg = reader.next_message().unwrap();
+        check_alert_msg(&msg);
+
+        handle.join().unwrap();
```

Comment:
For these longer tests, could you add some comments for each step?

Reply:
Sure, done